File: talk.py

class Symbol:
	key = '#'
	author = '@'
	goto = '>'
	condition = '?'
	action = '!'
	comment = '%'
	left = '{'
	right = '}'

def prefixcount(string, prefix):
	i = 0
	while string[i] == prefix:
		i += 1
	return i

def extract(string, startsymbol, endsymbol):
	if not startsymbol in string and endsymbol in string:
		return ('', string)
	
	start = string.find(startsymbol) + len(startsymbol)
	end = -1
	depth = 0
	for i in range(start, len(string)):
		s = string[i:]
		if not depth and s.startswith(endsymbol):
			end = i
			break
		depth += s.startswith(Symbol.left)
		depth -= s.startswith(Symbol.right)
	
	if end == -1:
		return ('', string)
	
	extraction = string[start:end]
	return (extraction, string.replace(startsymbol + extraction + endsymbol, ''))

def extractattribute(string, symbol):
	return extract(string, symbol + Symbol.left, Symbol.right)

def extractkey(string):
	return extractattribute(string, Symbol.key)[0]

def isempty(string):
	return string == '' or string.isspace()

class Line:
	def __init__(self, key, text='', author='', goto='', choices=[], condition='', action='', comment=''):
		self.key = key
		self.text = text
		self.author = author
		self.goto = goto
		self.choices = choices
		self.condition = condition
		self.action = action
		self.comment = comment

	def __repr__(self) -> str:
		return self.author + ': ' + self.text

class Talk:
	def __init__(self):
		self.lines = {}
		self.key = None
	
	@staticmethod
	def fromString(talk_string):

		talk = Talk()

		talk_string = talk_string.replace('\\\n\t', '\\\n')
		talk_string = talk_string.replace('\\\n', '')
		
		lines = [line for line in talk_string.splitlines() if not isempty(line)]

		for i, line in enumerate(lines):
			if isempty(extractkey(line)):
				lines[i] = line + Symbol.key + Symbol.left + Symbol.key + str(i) + Symbol.right

		previousauthor = ''

		for i, line in enumerate(lines):
			choices = []
			goto, text = extractattribute(line, Symbol.goto)
			
			if isempty(goto):
				tabs_i = prefixcount(line, '\t')
				if tabs_i % 2 == 0:
					tabs_min = tabs_i
					for j in range(i + 1, len(lines)):
						tabs_j = prefixcount(lines[j], '\t')
						tabs_min = min(tabs_j, tabs_min)

						if tabs_j <= tabs_min:
							if choices:
								break
							if tabs_j % 2 == 0:
								goto = extractkey(lines[j])
								break
						
						if tabs_j == tabs_i + 1 and tabs_i == tabs_min:
							choices.append(extractkey(lines[j]))
				
				elif i + 1 < len(lines):
					goto = extractkey(lines[i + 1])

			key, text = extractattribute(text, Symbol.key)
			author, text = extractattribute(text, Symbol.author)
			condition, text = extractattribute(text, Symbol.condition)
			action, text = extractattribute(text, Symbol.action)
			comment, text = extractattribute(text, Symbol.comment)

			talk.addLine(Line(
				key,
				text = text.strip(),
				author = author if not isempty(author) else previousauthor,
				goto = goto,
				choices = choices,
				condition = condition,
				action = action,
				comment = comment
			))

			if i == 0:
				talk.key = key

			previousauthor = author
		
		return talk

	def addLine(self, line: Line):
		self.lines[line.key] = line
	
	def talk(self):
		string = ''
		line = self.lines[self.key]

		while isempty(line.text):
			self.key = line.goto
			line = self.lines[self.key]

		string += str(line) + '\n'

		choices = [l for l in self.lines.values() if l.key in line.choices]
		if choices:
			for i, choice in enumerate(choices):
				string += str(i) + ' ' + str(choice) + '\n'
		
		return string

	def input(self, string):
		line = self.lines[self.key]
		choices = [l.key for l in self.lines.values() if l.key in line.choices]
		
		if choices:
			try:
				self.key = choices[int(string)]
			except:
				self.key = choices[0]
		else:
			self.key = line.goto
		

File: javascript/src/talk.ts
import { TalkSymbol } from './symbol.js';
import {
  Extraction,
  extractAttribute,
  extractKey,
  isEmpty,
  prefixCount,
} from './util.js';

class Line {
  key: string;
  text: string;
  author: string;
  goto: string;
  choices: string[];
  condition: string;
  action: string;
  comment: string;

  constructor(
    key: string,
    text: string,
    author: string,
    goto: string,
    choices: string[],
    condition: string,
    action: string,
    comment: string
  ) {
    this.key = key;
    this.text = text;
    this.author = author;
    this.goto = goto;
    this.choices = choices;
    this.condition = condition;
    this.action = action;
    this.comment = comment;
  }

  toString() {
    return this.author.length > 0 ? this.author + ': ' + this.text : this.text;
  }

  getChoices(talk: Talk) {
    return [...talk.lines.values()].filter((l) => this.choices.includes(l.key));
  }
}

export class Talk {
  lines: Map<string, Line>;
  private key: string;

  constructor(lines: Map<string, Line>, key: string) {
    this.lines = lines;
    this.setKey(key);
    this.key = key;
  }

  static fromString(talk_string: string): Talk {
    talk_string = talk_string.replace('\\\n\t', '\\\n');
    talk_string = talk_string.replace('\\\n', '');
    const lines = talk_string.split('\n').filter((line) => !isEmpty(line));

    // Add key to lines without keys
    lines.forEach((line, i) => {
      if (isEmpty(extractKey(line))) {
        const key_attr =
          TalkSymbol.key +
          TalkSymbol.left +
          TalkSymbol.key +
          i +
          TalkSymbol.right;
        lines[i] = line + key_attr;
      }
    });

    const talkLines = new Map();
    let authorPrev = '';
    lines.forEach((line, i) => {
      const choices: string[] = [];
      let extraction: Extraction;
      extraction = extractAttribute(line, TalkSymbol.goto);
      let goto = extraction.extraction;

      if (isEmpty(goto)) {
        const tabsI = prefixCount(line, '\t');
        if (tabsI % 2 == 0) {
          let tabsMin = tabsI;
          for (let j = i + 1; j < lines.length; j++) {
            const tabsJ = prefixCount(lines[j], '\t');
            tabsMin = Math.min(tabsJ, tabsMin);
            if (tabsJ <= tabsMin) {
              if (choices.length > 0) {
                break;
              }
              if (tabsJ % 2 == 0) {
                goto = extractKey(lines[j]);
                break;
              }
            }
            if (tabsJ == tabsI + 1 && tabsI == tabsMin) {
              choices.push(extractKey(lines[j]));
            }
          }
        } else if (i + 1 < lines.length) {
          goto = extractKey(lines[i + 1]);
        }
      }
      extraction = extractAttribute(extraction.rest, TalkSymbol.key);
      const key = extraction.extraction;
      extraction = extractAttribute(extraction.rest, TalkSymbol.author);
      const author = !isEmpty(extraction.extraction)
        ? extraction.extraction
        : authorPrev;
      extraction = extractAttribute(extraction.rest, TalkSymbol.condition);
      const condition = extraction.extraction;
      extraction = extractAttribute(extraction.rest, TalkSymbol.action);
      const action = extraction.extraction;
      extraction = extractAttribute(extraction.rest, TalkSymbol.comment);
      const comment = extraction.extraction;

      talkLines.set(
        key,
        new Line(
          key,
          extraction.rest.trim(),
          author,
          goto,
          choices,
          condition,
          action,
          comment
        )
      );
      authorPrev = author;
    });

    return new Talk(talkLines, extractKey(lines[0]));
  }

  talk() {
    if (this.key == null) return 'ERROR: Key is null!';
    const line = this.lines.get(this.key);
    if (line == null) return 'ERROR: Invalid key!';

    let result = line.toString() + '\n';
    line.getChoices(this).forEach((choice, i) => {
      result += i + ' ' + choice + '\n';
    });

    return result;
  }

  input(string: string) {
    const line = this.lines.get(this.key);
    if (line == null) return 'ERROR: Invalid key!';
    const choices = line.getChoices(this);
    if (choices.length > 0) {
      try {
        this.setKey(choices[parseInt(string)].goto);
      } catch {
        this.setKey(choices[0].goto);
      }
    } else {
      this.setKey(line.goto);
    }
  }

  setKey(key: string) {
    if (!this.lines.has(key)) {
      throw new Error('Key is not valid!');
    }
    this.key = key;
  }
}
